Project/DAQ_System/Program/Src/Platform/CAN/STM/F4/bxcan_stmf4.hpp

```diff
@@ -96,8 +96,8 @@ class BxCanStmF4 : public ICan {
 private:
 
 	uint8_t rx_buffer_[kMaxBytes] = { 0 };
-	bool message_arrived_ = false;
-	uint32_t can_id_ = 0;
+	volatile bool message_arrived_ = false;
+	volatile uint32_t can_id_ = 0;
```

Comment:
There's multiple use cases for it, but overall, it's for cases we do **not** want the compiler to optimize reading from a variable.

Normally, to improve performance, the compiler may opt to cache a variable (that is, make another copy of it). So, instead of re-reading the actual register in memory, you may get the copy.

This is not desirable in cases where the variable is updated a lot — such as in Interrupt Sub Routines (ISR) callback functions.

`volatile` is fairly common in firmware, and we do not use it a lot ourselves because we rely on the HAL library. I think I will mention it in the wiki, but I won't go too in depth as it can be gleamed from research.